File: apps/scraper/query/sources/course_catalog.py
"""The ASU course catalog: the description, credit hours and prerequisites of a course."""

from __future__ import annotations

from scraper.core.types import QueryParam, QuerySource
from scraper.query.params import term_code, text, url


def to_url(params: dict[str, str]) -> str:
    """The course list, narrowed to a keyword and, when one was given, to a term."""
    term = text(params, "term")
    return url(
        "https://catalog.apps.asu.edu/catalog/courses/courselist",
        [
            ("keywords", text(params, "keywords")),
            ("term", term_code(term) if term else ""),
            ("advanced", "false"),
        ],
    )


QUERY = QuerySource(
    key="course_catalog",
    description=(
        "Look one ASU course up in the catalog: what it covers, how many credit hours it "
        "carries, and the prerequisites it lists. It holds no sections and no seat counts."
    ),
    params=(
        QueryParam("keywords", "Course code, subject or title.", required=True, example="CSE 485"),
        QueryParam("term", "Term the catalog entry is read for.", example="Fall 2026"),
    ),
    to_url=to_url,
    needs_js=True,
    category="course_catalog",
)

File: apps/engine/src/runtime/tools/knowledge/search/course_catalog.rs
//! What one course covers, its credit hours, and the prerequisites it lists.

use chrono::NaiveDate;

use super::courses::code_in;
use super::{LiveSource, Param};

/// The ASU course catalog.
pub struct CourseCatalog;

const PARAMS: &[Param] = &[
    Param::text("keywords", "Course code, subject or title.")
        .required()
        .example("CSE 485"),
    Param::text("term", "Term the catalog entry is read for.").example("Fall 2026"),
];

impl LiveSource for CourseCatalog {
    fn key(&self) -> &'static str {
        "course_catalog"
    }

    fn hint(&self) -> &'static str {
        "what a course covers, its credits and prerequisites"
    }

    fn label(&self) -> &'static str {
        "ASU Course Catalog"
    }

    fn category(&self) -> &'static str {
        "course_catalog"
    }

    fn params(&self) -> &'static [Param] {
        PARAMS
    }

    fn query_param(&self) -> Option<&'static str> {
        Some("keywords")
    }

    fn narrow(&self, query: &str) -> String {
        code_in(query).unwrap_or_else(|| query.trim().to_owned())
    }

    fn derived(&self, query: &str, _today: NaiveDate) -> Vec<(&'static str, String)> {
        match super::courses::term_in(query) {
            Some(term) => vec![("term", term)],
            None => Vec::new(),
        }
    }
}
